pkg/bytecode/vm.go

```diff
@@ -63,11 +71,50 @@ func (vm *VM) Run() error {
 			globalIndex := ReadUint16(vm.instructions[ip+1:])
 			ip += 2
 			vm.globals[globalIndex] = vm.pop()
+		case OpAdd, OpSubtract:
```

Comment:
Since, as @juliaogris said, the type checking of the args should have been done by the parser, you could optimise all this as:
```
case OpAdd:
    right := vm.pop().(*numVal).V
    left := vm.pop().(*numVal).V
    vm.push(&numVal{V: left+right})
case OpSubtract:
    right := vm.pop().(*numVal).V
    left := vm.pop().(*numVal).V
    vm.push(&numVal{V: left-right})
```
For extra safety, you could define `vm.popNumVal()` that panics with a better error than the type assertion would (and probably good to make it panic with an `error` that `Is` some sort of `ErrInternal`).

Another comment I was going to make last PR that is relevant here now is `numVal` could just be a `float64` and not a struct with one. Get rid of the pointer too and it is likely that the value itself will fit inside the interface pair. Get rid of all the `.V` stuff, perhaps add a `newNumVal` to at least abstract the representation, and it would make the code look like:
```
case OpAdd:
    right := vm.popNum()
    left := vm.popNum()
    vm.push(newNum(left+right))
case OpSubtract:
    right := vm.popNum()
    left := vm.popNum()
    vm.push(newNum(left-right))
```

And since binary ops are rather common, you could define `popBinaryNums()` that is optimised as a single function call so you can do:
```
case OpAdd:
    right, left := vm.popBinaryNums()
    vm.push(newNum(left+right))
case OpSubtract:
    right, left := vm.popBinaryNums()
    vm.push(newNum(left-right))
```

Now we're getting tight on the hot path.

Reply:
> Get rid of the pointer too and it is likely that the value itself will fit inside the interface pair.

I think I'm wrong about this. I think earlier versions of Go might have stored values in the `data` field of an `iface`, if they fit, but I don't think they do any more (in fact I can't find any information on that at all historically or not, so now I think I'm imagining it. Looks like a non-pointer won't save anything anyway. That said, I still think `float64` is a better underlying type for `numVal`, but if it has a constructor (and I've locally added an `asFloat64()` method) then that can isolate the representation, which might give you some flexibility in optimizing later..

Reply:
It would be great if you could elaborate on the advantages that these changes would bring to the implementation, I see some downsides:

- drift with evaluator value (I still think it's a good idea to unify these in an abi package)
- vm is not protected against changes to the parser, e.g. if we ever decide that you can concatenate a num and a string this code will panic in a confusing manner instead of returning a proper error
- I think the code above becomes more complicated than the existing when string concatenation and boolean expressions are added

Reply:
Let's talk briefly about error checking / validation.

In general, in my view, errors should only be checked once at the boundaries of a "component". To me the "component" is the Evy toolchain.

If you feel strongly that parser Errors should be confirmed again in the `bytecode`
package then they should only be checked at (bytecode) compile time and not at (bytecode) execution time, because:

* Execution is as you called it the "hot path", don't do avoidable steps there.
* Don't duplicate checks for readability (less code is more readable especially to beginners).
* Don't duplicate checks because of `evy.wasm` size concerns. In general we want to avoid sending evy toolchain bytes over the network to the evy.dev site if they can be avoided.

To respond to your points in specific:

* **drift with evaluator value**: I will wear all consolidation efforts. TBH the evaluator.numVal etc. should just be a float 64 and not a struct wrapping a float64, my bad, this was an oversight. I have raise an [evaluator cleanup issue](https://github.com/evylang/todo/issues/66).
* **vm is not protected against changes to the parser**: The parser and vm are delivered together, if the language changes everything needs to be update. We are writing to the evy [spec](https://github.com/evylang/evy/blob/main/docs/spec.md#operators-and-expressions), let's not worry about possible future changes. 
If you insist on type checks in the bytecode package they should happen in the compile step not in the hotloop of the execution.
* **I think the code above becomes more complicated than the existing when string concatenation and boolean expressions are added**

So I've followed along and written it _all_ out as Cam suggested and to me this reads direct and obvious - I don't need to hop to helper functions to understand what's going on.

```go
case OpAdd:
    right, left := vm.popBinaryNums()
    vm.push(newNum(left+right))
case OpSubtract:
    right, left := vm.popBinaryNums()
    vm.push(newNum(left-right))
case OpMulitply: //...
case OpDivide: //...
case OpModulo: //...
case OpNumEQ:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left == right))
case OpNumNEQ:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left != right))
case OpNumLT:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left < right))
case OpNumLTE:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left <= right))
case OpNumGT:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left > right))
case OpNumGTE:
    right, left := vm.popBinaryNums()
    vm.push(newBool(left >= right))
case OpStringConcat:
    right, left := vm.popBinaryString()
    vm.push(newString(left+right))
case OpStringEQ:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left == right))
case OpStringNEQ:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left != right))
case OpStringLT:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left < right))
case OpStringLTE:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left <= right))
case OpStringGT:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left > right))
case OpStringGTE:
    right, left := vm.popBinaryStrings()
    vm.push(newBool(left >= right))
case OpArrayConcat:
    right, left := vm.popBinaryArray()
    vm.push(newArray(left+right))
case OpArrayEQ: 
    right, left := vm.popBinaryArray()
    vm.push(newBool(left.Equals(right)))
case OpArrayNEQ: 
    right, left := vm.popBinaryArray()
    vm.push(newBool(!left.Equals(right)))
case OpMapEQ: 
    right, left := vm.popBinaryArray()
    vm.push(newBool(left.Equals(right)))
case OpMapNEQ: 
    right, left := vm.popBinaryArray()
    vm.push(newBool(!left.Equals(right)))
case OpAnyEQ: // ...
case OpAnyNEQ: // ...
```

Reply:
Yeah, pretty much what Julia said. The advantages are largely about a shorter hot path, but also simpler code, something we've been continuously striving for with evy. We would like evy itself to be quite readable and approachable so that evy becomes an accessible code base to learners who are ready to move past the evy language itself.

I felt that:
```
case OpAddition:
    right, left := vm.popBinaryNums()
    vm.push(newNum(left + right))
```

explains the operation quite succinctly.

I don't think you need to be doing type validation in the VM - the parser has already validated that, but if you do feel it still needs to be done, I'd do it in the compiler. I'd front-load as much as I could into the compiler to keep the VM lean and fast. I'm also making the assumption that there are no plans to load external bytecode (i.e. it's not the JVM for example where source is compiled to bytecode and later executed). Instead we'll be compiling and executing in one hit, so you should be able to trust the validation of the parser or compiler. If my assumption is wrong, then perhaps we should talk about the far plans sometime, but even then I'd be considering other approaches to run-time type validation.

I would also not be using the `OpAddition` opcode/operation to be doing string concatenation or array concatenation - I would have the compiler identify those operations from the AST and have separate ops for those two operations. Keep the opcodes simple like CPU instructions are. Don't decide at runtime one of many meanings of an opcode, instead decide at compile time and have many opcodes.

With regard to the ABI, the point I ended up at was that we don't yet know enough what that should look like. I would very much prefer you to write your bytecode code for itself and when that is done, we can see what might apply to an ABI. I think we're all learning here, so I prefer not to look too far ahead because its so easy to get that wrong without experience.

Reply:
Thanks for all the feedback, I have incorporated these changes and will bear the style in mind as I continue 🙏